File: 06-generics/src/main.c
#include <stdio.h>
#include <math.h>

typedef struct {
    void *arg;
    float (*area)(void*);
    const char* (*name)(void);
} shape_t;

typedef struct {
    float radius;
} circle_t;

float circle_area(void *arg) {
    circle_t *circle = arg;

    return M_PI * circle->radius * circle->radius;
}

const char *circle_name() {
    return "circle";
}

typedef struct {
    float width;
    float height;
} rectangle_t;

float rectangle_area(void *arg) {
    rectangle_t *rectangle = arg;

    return rectangle->width * rectangle->height;
}

const char *rectangle_name() {
    return "rectangle";
}

typedef struct {
    float base;
    float height;
} triangle_t;

float triangle_area(void *arg) {
    triangle_t *triangle = arg;

    return (triangle->base * triangle->height) / 2.0;
}

const char *triangle_name() {
    return "triangle";
}

void print_shape(shape_t *shape) {
    printf("Shape is a %s with an area of %f\n",
	   shape->name(), shape->area(shape->arg));

}

int main(void) {
    circle_t circle = {
	.radius = 4.5,
    };
    shape_t circle_shape = {
	.arg = &circle,
	.area = circle_area,
	.name = circle_name,
    };

    rectangle_t rectangle = {
	.width = 3.6,
	.height = 6.8,
    };
    shape_t rectangle_shape = {
	.arg = &rectangle,
	.area = rectangle_area,
	.name = rectangle_name,
    };

    triangle_t triangle = {
	.base = 5.3,
	.height = 2.3,
    };
    shape_t triangle_shape = {
	.arg = &triangle,
	.area = triangle_area,
	.name = triangle_name,
    };

    print_shape(&circle_shape);
    print_shape(&rectangle_shape);
    print_shape(&triangle_shape);

    return 0;
}

File: 06-generics/src/main.rs
use core::f32;

trait Shape {
    fn area(&self) -> f32;
    fn name() -> &'static str;
}

struct Circle {
    radius: f32,
}

impl Shape for Circle {
    fn area(&self) -> f32 {
        f32::consts::PI * self.radius * self.radius
    }

    fn name() -> &'static str {
        "circle"
    }
}

struct Rectangle {
    width: f32,
    height: f32,
}

impl Shape for Rectangle {
    fn area(&self) -> f32 {
        self.width * self.height
    }

    fn name() -> &'static str {
        "rectangle"
    }
}

struct Triangle {
    base: f32,
    height: f32,
}

impl Shape for Triangle {
    fn area(&self) -> f32 {
        (self.base * self.height) / 2.0
    }

    fn name() -> &'static str {
        "triangle"
    }
}

fn print_shape<T: Shape>(shape: &T) {
    println!("Shape is a {} with an area of {}",
	     T::name(), shape.area());
}

fn main() {
    let circle = Circle {
        radius: 4.5,
    };
    let rectangle = Rectangle {
	width: 3.6,
	height: 6.8,
    };
    let triangle = Triangle {
	base: 5.3,
	height: 2.3,
    };

    print_shape(&circle);
    print_shape(&rectangle);
    print_shape(&triangle);
}
